### PR TITLE
Go SDK transform support

### PR DESCRIPTION
**Pull request by Frassle**
# Description

This adds a new experimental feature to the Go SDK to register remote transform functions. These are currently all prefixed 'X' to show they're experimental (they can't be in their own package because of circular dependencies). 

These transform functions will run even for resources created inside MLCs.

## Checklist

- [x] I have run `make tidy` to update any new dependencies
- [x] I have run `make lint` to verify my code passes the lint check
  - [x] I have formatted my code using `gofumpt`

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

**Comment by Frassle** [comment]
Still need to fix up the mock support for this, but ready to start looking at for the rest of the changes.

File: sdk/go/pulumi/context.go [hunk 1]
```diff
@@ -50,18 +51,21 @@ var disableResourceReferences = cmdutil.IsTruthy(os.Getenv("PULUMI_DISABLE_RESOU
...
+	engine        pulumirpc.EngineClient
+	engineConn    *grpc.ClientConn
+	callbacksLock sync.Mutex
+	callbacks     *callbackServer
 
 	keepResources       bool       // true if resources should be marshaled as strongly-typed references.
 	keepOutputValues    bool       // true if outputs should be marshaled as strongly-type output values.
 	supportsDeletedWith bool       // true if deletedWith supported by pulumi
 	supportsAliasSpecs  bool       // true if full alias specification is supported by pulumi
+	supportsTransforms  bool       // true if remote transformations are supported by pulumi
```
Comment (justinvp) [inline sdk/go/pulumi/context.go:68]: Nit:

```suggestion
	supportsTransforms  bool       // true if remote transforms are supported by pulumi
```

File: sdk/go/pulumi/context.go [hunk 2]
```diff
@@ -282,6 +293,230 @@ func (ctx *Context) IsConfigSecret(key string) bool {
 	return false
 }
 
+// registerTransform starts up a callback server if not already running and registers the given transformation.
```
Comment (justinvp) [inline sdk/go/pulumi/context.go:296]:
```suggestion
// registerTransform starts up a callback server if not already running and registers the given transform.
```

File: sdk/go/pulumi/context.go [hunk 3]
```diff
@@ -282,6 +293,230 @@ func (ctx *Context) IsConfigSecret(key string) bool {
...
+func (ctx *Context) registerTransform(t XResourceTransform) (*pulumirpc.Callback, error) {
+	contract.Assertf(ctx.state.supportsTransforms, "transformation attempted but not supported by resource monitor")
```
Comment (justinvp) [inline sdk/go/pulumi/context.go:298]:
```suggestion
	contract.Assertf(ctx.state.supportsTransforms, "transform attempted but not supported by resource monitor")
```

File: sdk/go/pulumi/context.go [hunk 4]
```diff
@@ -282,6 +293,230 @@ func (ctx *Context) IsConfigSecret(key string) bool {
...
+
+	// Wrap the transformation in a callback function.
```
Comment (justinvp) [inline sdk/go/pulumi/context.go:300]:
```suggestion
	// Wrap the transform in a callback function.
```

File: sdk/go/pulumi/context.go [hunk 5]
```diff
@@ -282,6 +293,230 @@ func (ctx *Context) IsConfigSecret(key string) bool {
...
+	callback := func(innerCtx context.Context, req []byte) (proto.Message, error) {
+		// Unmarshal the request to a pulumirpc.ResourceTransformationRequest.
```
Comment (justinvp) [inline sdk/go/pulumi/context.go:302]:
```suggestion
		// Unmarshal the request to a pulumirpc.TransformRequest.
```

File: sdk/go/pulumi/context.go [hunk 6]
```diff
@@ -635,9 +870,13 @@ func (ctx *Context) ReadResource(
...
-	if err != nil {
-		return err
+	var transformations []ResourceTransformation
+	if !ctx.state.supportsTransforms {
+		var err error
+		props, options, transformations, err = applyTransformations(t, name, props, resource, opts, options)
+		if err != nil {
+			return err
+		}
 	}
```
Comment (justinvp) [inline sdk/go/pulumi/context.go:880]: We want these to continue to work, right?

File: sdk/go/pulumi/context.go [hunk 7]
```diff
@@ -1851,6 +2090,23 @@ func (ctx *Context) RegisterStackTransformation(t ResourceTransformation) error
 	return nil
 }
 
+// XRegisterStackTransform adds a transform to all future resources constructed in this Pulumi stack.
+//
+// Experimental.
+func (ctx *Context) XRegisterStackTransform(t XResourceTransform) error {
+	if !ctx.state.supportsTransforms {
+		return fmt.Errorf("update engine")
```
Comment (justinvp) [inline sdk/go/pulumi/context.go:2098]: Should we have a nicer error message here? Something like "upgrade your version of the Pulumi CLI". We have other errors elsewhere (e.g. in automation API) that we could use as the basis for what we put here.

File: sdk/go/pulumi/rpc.go [hunk 8]
```diff
@@ -596,6 +597,135 @@ func unmarshalPropertyValue(ctx *Context, v resource.PropertyValue) (interface{}
...
+				}
+				return NewAssetArchive(as), nil
+			case archive.IsPath():
+				return NewFileArchive(archive.Path), nil
+			case archive.IsURI():
+				return NewRemoteArchive(archive.URI), nil
+			}
+			return nil, errors.New("expected archive to be one of Assets, File, or Remote; got none")
+		case v.IsResourceReference():
+			return nil, errors.New("TODO")
```
Comment (justinvp) [inline sdk/go/pulumi/rpc.go:667]: TODO?

File: tests/integration/backend/filestate/backend_azure_test.go
Comment (justinvp) [inline tests/integration/backend/filestate/backend_azure_test.go:1]: This file and others look unrelated

File: tests/integration/integration_nodejs_test.go
Comment (justinvp) [inline tests/integration/integration_nodejs_test.go:1]: Looks unrelated

File: tests/integration/integration_test.go
Comment (justinvp) [inline tests/integration/integration_test.go:1]: unrelated

File: tests/integration/transformations/transformations_nodejs_test.go [hunk 9]
```diff
@@ -25,6 +25,8 @@ func TestNodejsTransformations(t *testing.T) {
 				},
 				Quick:                  true,
 				ExtraRuntimeValidation: Validator,
+				DebugLogLevel:          10,
+				Verbose:                true,
```
Comment (justinvp) [inline tests/integration/transformations/transformations_nodejs_test.go:29]: Is this needed for this PR?

File: tests/integration/backend/filestate/backend_azure_test.go
Comment (Frassle) [reply tests/integration/backend/filestate/backend_azure_test.go:1]: hmm yeh bad merge I think, I'll get them removed

File: sdk/go/pulumi/context.go [hunk 10]
```diff
@@ -282,6 +293,229 @@ func (ctx *Context) IsConfigSecret(key string) bool {
 	return false
 }
 
+// registerTransform starts up a callback server if not already running and registers the given transform.
+func (ctx *Context) registerTransform(t XResourceTransform) (*pulumirpc.Callback, error) {
+	contract.Assertf(ctx.state.supportsTransforms, "transform attempted but not supported by resource monitor")
```
Comment (tgummerer) [inline sdk/go/pulumi/context.go:298]: Could this just be a regular error rather than an assert?  I the call [here](https://github.com/pulumi/pulumi/pull/15355/files#diff-6ac1847413cb55c8d480423da6e23ed6ac98c0066fcf7c57f8b5cd960a04dee1R2106) is protected by an error return, but could we end up [here](https://github.com/pulumi/pulumi/pull/15355/files#diff-6ac1847413cb55c8d480423da6e23ed6ac98c0066fcf7c57f8b5cd960a04dee1R1161) and run into this assert?  Or maybe that's not possible and I'm just not familiar enough with the engine yet?

File: sdk/go/pulumi/resource.go [hunk 11]
```diff
@@ -369,6 +379,12 @@ type ResourceOptions struct {
 	// the resource's properties during construction.
 	Transformations []ResourceTransformation
 
+	// XTransforms is a list of functions that transform
+	// the resource's properties during construction.
+	//
+	// Experimental.
```
Comment (tgummerer) [inline sdk/go/pulumi/resource.go:385]: Is it worth explaining a bit more what "Experimental" entails?  E.g. API changes are expected, behaviour changes are expected, etc.

File: tests/integration/transforms/go/simple/main.go [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+			},
+		}))
+		if err != nil {
+			return err
+		}
+
+		// Scenario #3 - apply a transform to the Stack to transform all (future) resources in the stack
+		err = ctx.XRegisterStackTransform(func(rta *pulumi.XResourceTransformArgs) *pulumi.XResourceTransformResult {
+			fmt.Printf("stack transform\n")
+			fmt.Printf("%v %v\n", rta.Type, rta.Props)
```
Comment (tgummerer) [inline tests/integration/transforms/go/simple/main.go:106]: debugging leftovers?

File: tests/integration/transforms/go/simple/main.go [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+
+					return &pulumi.XResourceTransformResult{
+						Props: rta.Props,
+						Opts:  rta.Opts,
+					}
+				}
+				return nil
+			},
+			func(rta *pulumi.XResourceTransformArgs) *pulumi.XResourceTransformResult {
+				fmt.Printf("res4 transform 2\n")
```
Comment (tgummerer) [inline tests/integration/transforms/go/simple/main.go:148]: debugging leftovers?

File: tests/integration/transforms/go/simple/random.go [hunk 14]
```diff
@@ -0,0 +1,49 @@
+// Copyright 2016-2021, Pulumi Corporation.  All rights reserved.
```
Comment (tgummerer) [inline tests/integration/transforms/go/simple/random.go:1]:
```suggestion
// Copyright 2016-2024, Pulumi Corporation.  All rights reserved.
```

File: tests/integration/transforms/go/simple/main.go [hunk 15]
```diff
@@ -0,0 +1,186 @@
+// Copyright 2016-2023, Pulumi Corporation.  All rights reserved.
```
Comment (tgummerer) [inline tests/integration/transforms/go/simple/main.go:1]:
```suggestion
// Copyright 2016-2024, Pulumi Corporation.  All rights reserved.
```

File: tests/integration/transforms/transforms_go_test.go [hunk 16]
```diff
@@ -0,0 +1,33 @@
+// Copyright 2016-2023, Pulumi Corporation.  All rights reserved.
```
Comment (tgummerer) [inline tests/integration/transforms/transforms_go_test.go:1]:
```suggestion
// Copyright 2016-2024, Pulumi Corporation.  All rights reserved.
```

File: tests/integration/transforms/transforms_test.go [hunk 17]
```diff
@@ -0,0 +1,84 @@
+// Copyright 2016-2018, Pulumi Corporation.  All rights reserved.
```
Comment (tgummerer) [inline tests/integration/transforms/transforms_test.go:1]:
```suggestion
// Copyright 2016-2024, Pulumi Corporation.  All rights reserved.
```

File: tests/integration/transforms/go/simple/main.go [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+			},
+		}))
+		if err != nil {
+			return err
+		}
+
+		// Scenario #5 - mutate the properties of a resource
+		_, err = NewRandom(ctx, "res5", &RandomArgs{Length: pulumi.Int(10)}, pulumi.XTransforms([]pulumi.XResourceTransform{
+			func(rta *pulumi.XResourceTransformArgs) *pulumi.XResourceTransformResult {
+				fmt.Printf("res5 transform\n")
```
Comment (tgummerer) [inline tests/integration/transforms/go/simple/main.go:167]: more debugging leftovers?

File: tests/integration/transforms/go/simple/main.go [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+			},
+		}))
+		if err != nil {
+			return err
+		}
+
+		// Scenario #2 - apply a transform to a Component to transform it's children
+		_, err = NewMyComponent(ctx, "res2", pulumi.XTransforms([]pulumi.XResourceTransform{
+			func(rta *pulumi.XResourceTransformArgs) *pulumi.XResourceTransformResult {
+				fmt.Printf("res2 transform\n")
```
Comment (tgummerer) [inline tests/integration/transforms/go/simple/main.go:88]: debugging leftovers?

File: tests/integration/transforms/go/simple/main.go [hunk 20]
```diff
@@ -0,0 +1,186 @@
...
+	component.Child2 = child2
+	return component, nil
+}
+
+func main() {
+	pulumi.Run(func(ctx *pulumi.Context) error {
+		// Scenario #1 - apply a transform to a CustomResource
+		_, err := NewRandom(ctx, "res1", &RandomArgs{Length: pulumi.Int(5)}, pulumi.XTransforms([]pulumi.XResourceTransform{
+			func(rta *pulumi.XResourceTransformArgs) *pulumi.XResourceTransformResult {
+				fmt.Printf("res1 transform\n")
```
Comment (tgummerer) [inline tests/integration/transforms/go/simple/main.go:73]: debugging leftovers?

File: tests/integration/transforms/go/simple/main.go (hunk 18 above)
Comment (Frassle) [reply tests/integration/transforms/go/simple/main.go:167]: Kinda, the transform test I copied for this had print statements left in as well.

File: sdk/go/pulumi/context.go (hunk 10 above)
Comment (Frassle) [reply sdk/go/pulumi/context.go]: Yup moved the check

File: sdk/go/pulumi/resource.go (hunk 11 above)
Comment (Frassle) [reply sdk/go/pulumi/resource.go:385]: We should probably just have a doc somewhere on the site to explain it and link to that, rather than repeating it everywhere. But I imagine most people are ok with what experimental means.

**Review by tgummerer — APPROVED** [review]
